File: backend/src/test/java/org/luxons/sevenwonders/lobby/LobbyTest.java
package org.luxons.sevenwonders.lobby;

import java.util.Arrays;

import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.luxons.sevenwonders.game.Game;
import org.luxons.sevenwonders.game.api.CustomizableSettings;
import org.luxons.sevenwonders.game.data.GameDefinition;
import org.luxons.sevenwonders.game.data.GameDefinitionLoader;
import org.luxons.sevenwonders.lobby.Lobby.GameAlreadyStartedException;
import org.luxons.sevenwonders.lobby.Lobby.PlayerNameAlreadyUsedException;
import org.luxons.sevenwonders.lobby.Lobby.PlayerOverflowException;
import org.luxons.sevenwonders.lobby.Lobby.PlayerUnderflowException;
import org.luxons.sevenwonders.lobby.Lobby.UnknownPlayerException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.junit.Assume.assumeTrue;

@RunWith(Theories.class)
public class LobbyTest {

    @DataPoints
    public static int[] nbPlayers() {
        return new int[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    }

    @Rule
    public ExpectedException thrown = ExpectedException.none();

    private static GameDefinition gameDefinition;

    private Player gameOwner;

    private Lobby lobby;

    @BeforeClass
    public static void loadDefinition() {
        gameDefinition = new GameDefinitionLoader().getGameDefinition();
    }

    @Before
    public void setUp() {
        gameOwner = new Player("gameowner", "Game owner");
        lobby = new Lobby(0, "Test Game", gameOwner, gameDefinition);
    }

    @Test
    public void testId() {
        Lobby l = new Lobby(5, "Test Game", gameOwner, gameDefinition);
        assertEquals(5, l.getId());
    }

    @Test
    public void testName() {
        Lobby l = new Lobby(5, "Test Game", gameOwner, gameDefinition);
        assertEquals("Test Game", l.getName());
    }

    @Test
    public void testOwner() {
        gameOwner.setIndex(42);
        Lobby l = new Lobby(5, "Test Game", gameOwner, gameDefinition);
        assertSame(gameOwner, l.getPlayers().get(0));
        assertSame(l, gameOwner.getLobby());
        assertEquals(0, gameOwner.getIndex());
    }

    @Test
    public void isOwner_falseWhenNull() {
        assertFalse(lobby.isOwner(null));
    }

    @Test
    public void isOwner_falseWhenEmptyString() {
        assertFalse(lobby.isOwner(""));
    }

    @Test
    public void isOwner_falseWhenGarbageString() {
        assertFalse(lobby.isOwner("this is garbage"));
    }

    @Test
    public void isOwner_trueWhenOwnerUsername() {
        assertTrue(lobby.isOwner(gameOwner.getUsername()));
    }

    @Test
    public void isOwner_falseWhenOtherPlayerName() {
        Player player = new Player("testuser", "Test User");
        lobby.addPlayer(player);
        assertFalse(lobby.isOwner(player.getUsername()));
    }

    @Test
    public void addPlayer_success() {
        Player player = new Player("testuser", "Test User");
        lobby.addPlayer(player);
        assertTrue(lobby.containsUser("testuser"));
        assertSame(lobby, player.getLobby());
        assertSame(1, player.getIndex()); // the owner is 0
    }

    @Test(expected = PlayerNameAlreadyUsedException.class)
    public void addPlayer_failsOnSameName() {
        Player player = new Player("testuser", "Test User");
        Player player2 = new Player("testuser2", "Test User");
        lobby.addPlayer(player);
        lobby.addPlayer(player2);
    }

    @Test(expected = PlayerOverflowException.class)
    public void addPlayer_playerOverflowWhenTooMany() {
        // the owner + the max number gives an overflow
        addPlayers(gameDefinition.getMaxPlayers());
    }

    @Test(expected = GameAlreadyStartedException.class)
    public void addPlayer_failWhenGameStarted() {
        // total with owner is the minimum
        addPlayers(gameDefinition.getMinPlayers() - 1);
        lobby.startGame();
        lobby.addPlayer(new Player("soonerNextTime", "The Late Guy"));
    }

    private void addPlayers(int nbPlayers) {
        for (int i = 0; i < nbPlayers; i++) {
            Player player = new Player("testuser" + i, "Test User " + i);
            lobby.addPlayer(player);
        }
    }

    @Test(expected = UnknownPlayerException.class)
    public void removePlayer_failsWhenNotPresent() {
        lobby.removePlayer("anyname");
    }

    @Test
    public void removePlayer_success() {
        Player player = new Player("testuser", "Test User");
        lobby.addPlayer(player);
        lobby.removePlayer("testuser");
        assertFalse(lobby.containsUser("testuser"));
        assertNull(player.getLobby());
        assertNull(player.getGame());
        assertEquals(-1, player.getIndex());
    }

    @Test
    public void reorderPlayers_success() {
        Player player1 = new Player("testuser1", "Test User 1");
        Player player2 = new Player("testuser2", "Test User 2");
        Player player3 = new Player("testuser3", "Test User 3");
        lobby.addPlayer(player1);
        lobby.addPlayer(player2);
        lobby.addPlayer(player3);
        lobby.reorderPlayers(Arrays.asList("testuser3", "testuser1", "testuser2"));
        assertEquals("testuser3", lobby.getPlayers().get(0).getUsername());
        assertEquals("testuser1", lobby.getPlayers().get(1).getUsername());
        assertEquals("testuser2", lobby.getPlayers().get(2).getUsername());
        assertEquals(0, lobby.getPlayers().get(0).getIndex());
        assertEquals(1, lobby.getPlayers().get(1).getIndex());
        assertEquals(2, lobby.getPlayers().get(2).getIndex());
    }

    @Test(expected = UnknownPlayerException.class)
    public void reorderPlayers_failsOnUnknownPlayer() {
        Player player1 = new Player("testuser1", "Test User 1");
        Player player2 = new Player("testuser2", "Test User 2");
        Player player3 = new Player("testuser3", "Test User 3");
        lobby.addPlayer(player1);
        lobby.addPlayer(player2);
        lobby.addPlayer(player3);
        lobby.reorderPlayers(Arrays.asList("unknown", "testuser1", "testuser2"));
    }

    @Theory
    public void startGame_failsBelowMinPlayers(int nbPlayers) {
        assumeTrue(nbPlayers < gameDefinition.getMinPlayers());
        thrown.expect(PlayerUnderflowException.class);
        // there is already the owner
        addPlayers(nbPlayers - 1);
        lobby.startGame();
    }

    @Theory
    public void startGame_succeedsAboveMinPlayers(int nbPlayers) {
        assumeTrue(nbPlayers >= gameDefinition.getMinPlayers());
        assumeTrue(nbPlayers <= gameDefinition.getMaxPlayers());
        // there is already the owner
        addPlayers(nbPlayers - 1);
        Game game = lobby.startGame();
        assertNotNull(game);
        lobby.getPlayers().forEach(p -> assertSame(game, p.getGame()));
    }

    @Test
    public void startGame_switchesState() {
        assertTrue(lobby.getState() == State.LOBBY);
        // there is already the owner
        addPlayers(gameDefinition.getMinPlayers() - 1);
        lobby.startGame();
        assertTrue(lobby.getState() == State.PLAYING);
    }

    @Test
    public void setSettings() throws Exception {
        CustomizableSettings settings = new CustomizableSettings();
        lobby.setSettings(settings);
        assertSame(settings, lobby.getSettings());
    }
}

File: backend/src/test/kotlin/org/luxons/sevenwonders/lobby/LobbyTest.kt
package org.luxons.sevenwonders.lobby

import org.junit.Assert.*
import org.junit.Assume.assumeTrue
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Rule
import org.junit.Test
import org.junit.experimental.theories.DataPoints
import org.junit.experimental.theories.Theories
import org.junit.experimental.theories.Theory
import org.junit.rules.ExpectedException
import org.junit.runner.RunWith
import org.luxons.sevenwonders.game.api.CustomizableSettings
import org.luxons.sevenwonders.game.data.GameDefinition
import org.luxons.sevenwonders.game.data.GameDefinitionLoader
import org.luxons.sevenwonders.lobby.Lobby.*
import java.util.Arrays

@RunWith(Theories::class)
class LobbyTest {

    @JvmField
    @Rule
    var thrown = ExpectedException.none()

    private lateinit var gameOwner: Player

    private lateinit var lobby: Lobby

    @Before
    fun setUp() {
        gameOwner = Player("gameowner", "Game owner")
        lobby = Lobby(0, "Test Game", gameOwner, gameDefinition)
    }

    @Test
    fun testId() {
        val lobby = Lobby(5, "Test Game", gameOwner, gameDefinition)
        assertEquals(5, lobby.id)
    }

    @Test
    fun testName() {
        val lobby = Lobby(5, "Test Game", gameOwner, gameDefinition)
        assertEquals("Test Game", lobby.name)
    }

    @Test
    fun testOwner() {
        val lobby = Lobby(5, "Test Game", gameOwner, gameDefinition)
        assertSame(gameOwner, lobby.getPlayers()[0])
        assertSame(lobby, gameOwner.lobby)
    }

    @Test
    fun isOwner_falseWhenNull() {
        assertFalse(lobby.isOwner(null))
    }

    @Test
    fun isOwner_falseWhenEmptyString() {
        assertFalse(lobby.isOwner(""))
    }

    @Test
    fun isOwner_falseWhenGarbageString() {
        assertFalse(lobby.isOwner("this is garbage"))
    }

    @Test
    fun isOwner_trueWhenOwnerUsername() {
        assertTrue(lobby.isOwner(gameOwner.username))
    }

    @Test
    fun isOwner_falseWhenOtherPlayerName() {
        val player = Player("testuser", "Test User")
        lobby.addPlayer(player)
        assertFalse(lobby.isOwner(player.username))
    }

    @Test
    fun addPlayer_success() {
        val player = Player("testuser", "Test User")
        lobby.addPlayer(player)
        assertTrue(lobby.containsUser("testuser"))
        assertSame(lobby, player.lobby)
    }

    @Test(expected = PlayerNameAlreadyUsedException::class)
    fun addPlayer_failsOnSameName() {
        val player = Player("testuser", "Test User")
        val player2 = Player("testuser2", "Test User")
        lobby.addPlayer(player)
        lobby.addPlayer(player2)
    }

    @Test(expected = PlayerOverflowException::class)
    fun addPlayer_playerOverflowWhenTooMany() {
        // the owner + the max number gives an overflow
        addPlayers(gameDefinition.maxPlayers)
    }

    @Test(expected = GameAlreadyStartedException::class)
    fun addPlayer_failWhenGameStarted() {
        // total with owner is the minimum
        addPlayers(gameDefinition.minPlayers - 1)
        lobby.startGame()
        lobby.addPlayer(Player("soonerNextTime", "The Late Guy"))
    }

    private fun addPlayers(nbPlayers: Int) {
        for (i in 0 until nbPlayers) {
            val player = Player("testuser$i", "Test User $i")
            lobby.addPlayer(player)
        }
    }

    @Test(expected = UnknownPlayerException::class)
    fun removePlayer_failsWhenNotPresent() {
        lobby.removePlayer("anyname")
    }

    @Test
    fun removePlayer_success() {
        val player = Player("testuser", "Test User")
        lobby.addPlayer(player)
        assertTrue(player.isInLobby)
        assertFalse(player.isInGame)

        lobby.removePlayer("testuser")
        assertFalse(lobby.containsUser("testuser"))
        assertFalse(player.isInLobby)
        assertFalse(player.isInGame)
    }

    @Test
    fun reorderPlayers_success() {
        val player1 = Player("testuser1", "Test User 1")
        val player2 = Player("testuser2", "Test User 2")
        val player3 = Player("testuser3", "Test User 3")
        lobby.addPlayer(player1)
        lobby.addPlayer(player2)
        lobby.addPlayer(player3)
        lobby.reorderPlayers(Arrays.asList("testuser3", "testuser1", "testuser2"))
        assertEquals("testuser3", lobby.getPlayers()[0].username)
        assertEquals("testuser1", lobby.getPlayers()[1].username)
        assertEquals("testuser2", lobby.getPlayers()[2].username)
    }

    @Test(expected = UnknownPlayerException::class)
    fun reorderPlayers_failsOnUnknownPlayer() {
        val player1 = Player("testuser1", "Test User 1")
        val player2 = Player("testuser2", "Test User 2")
        val player3 = Player("testuser3", "Test User 3")
        lobby.addPlayer(player1)
        lobby.addPlayer(player2)
        lobby.addPlayer(player3)
        lobby.reorderPlayers(Arrays.asList("unknown", "testuser1", "testuser2"))
    }

    @Theory
    fun startGame_failsBelowMinPlayers(nbPlayers: Int) {
        assumeTrue(nbPlayers < gameDefinition.minPlayers)
        thrown.expect(PlayerUnderflowException::class.java)
        // there is already the owner
        addPlayers(nbPlayers - 1)
        lobby.startGame()
    }

    @Theory
    fun startGame_succeedsAboveMinPlayers(nbPlayers: Int) {
        assumeTrue(nbPlayers >= gameDefinition.minPlayers)
        assumeTrue(nbPlayers <= gameDefinition.maxPlayers)
        // there is already the owner
        addPlayers(nbPlayers - 1)

        assertEquals(nbPlayers, lobby.getPlayers().size)
        lobby.getPlayers().forEach {
            assertSame(lobby, it.lobby)
            assertTrue(it.isInLobby)
            assertFalse(it.isInGame)
        }

        val game = lobby.startGame()
        assertNotNull(game)
        lobby.getPlayers().forEachIndexed { index, it ->
            assertSame(index, it.index)
            assertSame(lobby, it.lobby)
            assertSame(game, it.game)
            assertTrue(it.isInLobby)
            assertTrue(it.isInGame)
        }
    }

    @Test
    fun startGame_switchesState() {
        assertTrue(lobby.state === State.LOBBY)
        // there is already the owner
        addPlayers(gameDefinition.minPlayers - 1)
        lobby.startGame()
        assertTrue(lobby.state === State.PLAYING)
    }

    @Test
    fun setSettings() {
        val settings = CustomizableSettings()
        lobby.settings = settings
        assertSame(settings, lobby.settings)
    }

    companion object {

        private lateinit var gameDefinition: GameDefinition

        @JvmStatic
        @DataPoints
        fun nbPlayers(): IntArray {
            return intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
        }

        @JvmStatic
        @BeforeClass
        fun loadDefinition() {
            gameDefinition = GameDefinitionLoader.gameDefinition
        }
    }
}
